sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,340 @@
+		state.fetchAdd(1);
+		return cast(void*) cast(size_t) lock.count;
+	}
+
+	auto writer = runThread(takeExclusiveLock);
+	{
+		bool triedExclusiveLock() {
+			return lock.count >= SharedLock.Exclusive;
+		}
+
+		lock.mutex.lock();
+		scope(exit) lock.mutex.unlock();
+
+		lock.mutex.waitFor(triedExclusiveLock);
+		assert(lock.count == SharedLock.Exclusive + 1);
+		assert(state.load() == 1);
+	}
+
+	lock.sharedUnlock();
+	assert(lock.count == SharedLock.Exclusive);
+	void* ret;
+	pthread_join(writer, &ret);
+	assert(cast(size_t) ret == SharedLock.Exclusive);
+	assert(state.load() == 2);
+	lock.exclusiveUnlock();
+	assert(lock.count == 0);
+}
+
+unittest threadStressTest {
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	import d.sync.atomic;
+	shared SharedLock lock;
+	shared Atomic!uint numSharedLocks;
+	shared Atomic!uint numExclusiveLocks;
+	shared Mutex mutex;
+	bool runTest;
+
+	bool shouldRunTest() {
+		return runTest;
+	}
+
+	enum MaxExclusive = 5;
+	enum MaxShared = 50;
+
+	void* runExclusive() {
+		mutex.lock();
+		mutex.waitFor(shouldRunTest);
+		mutex.unlock();
+
+		foreach (i; 0 .. 10000) {
+			lock.exclusiveLock();
+			scope(exit) lock.exclusiveUnlock();
+			auto ne = numExclusiveLocks.fetchAdd(1);
+			auto ns = numSharedLocks.load();
+			assert(ne == 0);
+			assert(ns == 0);
+			import sys.posix.sched;
+			sched_yield();
+			ne = numExclusiveLocks.fetchSub(1);
+			assert(ne == 1);
+		}
+
+		return null;
+	}
+
+	void* runShared() {
+		mutex.lock();
+		mutex.waitFor(shouldRunTest);
+		mutex.unlock();
+
+		size_t highWater = 0;
+		foreach (i; 0 .. 10000) {
+			lock.sharedLock();
+			scope(exit) lock.sharedUnlock();
+
+			auto ne = numExclusiveLocks.load();
+			auto ns = numSharedLocks.fetchAdd(1);
+			assert(ne == 0);
+			assert(ns >= 0 && ns < MaxShared);
+			import sys.posix.sched;
+			sched_yield();
+			ns = numSharedLocks.fetchSub(1);
+			assert(ns > 0 && ns <= MaxShared);
+			if (ns > highWater) {
+				highWater = ns;
+			}
+		}
+
+		return cast(void*) highWater;
+	}
+
+	import core.stdc.pthread;
+	pthread_t[MaxExclusive] exclusives;
+	pthread_t[MaxShared] sharers;
+
+	foreach (i; 0 .. exclusives.length) {
+		exclusives[i] = runThread(runExclusive);
+	}
+
+	foreach (i; 0 .. sharers.length) {
+		sharers[i] = runThread(runShared);
+	}
+
+	mutex.lock();
+	runTest = true;
+	mutex.unlock();
+
+	foreach (i; 0 .. exclusives.length) {
+		void* ret;
+		pthread_join(exclusives[i], &ret);
+	}
+
+	size_t highWater;
+	foreach (i; 0 .. sharers.length) {
+		size_t ret;
+		pthread_join(sharers[i], cast(void**) &ret);
+		if (ret > highWater) {
+			highWater = ret;
+		}
+	}
+
+	import core.stdc.stdio;
+	printf("simultaneous sharers highWater = %lld\n", highWater);
```

Comment:
Is that important?

Reply:
It's a test to see how high the concurrent holders of the lock got. It's not important, and I don't think there's really an assert to do here, but I wanted to see the value.

It can be removed if needed, I did see other tests where things are printed instead of tested.